[![Snakemake](https://img.shields.io/badge/snakemake- >=3.8.0-brightgreen.svg?style=flat-square)](https://snakemake.bitbucket.io)

## Introduction
This NGS pipeline is designed for NCI Patient-Derived Models Repository (PDMR) project (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr)

## Installation

The easiest way for installation is to clone the repository using command below.
```
git clone https://github.com/FNL-MoCha/PDMR_landscape_paper_public.git
```
This pipeline is available on NIH biowulf cluster, contact us (NCI_PDM_Repository@mail.nih.gov or mochanctn@mail.nih.gov) if you would like to do a test run.

## Scope
### Inputs:
  - Raw sequencing reads (paired-end FASTQ files)
  - Required references (hg19 reference genome fasta)
  - Target BED files


### Outputs:
- QC reports (FastQC, MultiQC, coverage summaries)
- Aligned reads: coordinate-sorted BAM files with index and alignment metrics
- Variant calls:
	- Tumor-only: VCF file generated by HaplotypeCaller/ Platypus
	- Tumorâ€“normal pairs: VCF file generated by GATK Mutect2
- Copy-number profiles generated with CNVkit

## Prerequisites
- Linux environment (HPC or workstation)
- Bioinformatics Tools Listed in [config files](config/config_common.json)

## Conventions

- Sample names cannot have "/" or "." in them
- Fastq files end in ".fastq.gz"
- Fastq files are stored in DATA_DIR (Set as Environment Variable) 

### DNASeq:
----
- QC
- BWA
- Broad Standard Practices on bwa bam  
- Haplotype Caller, Platupys, MuTect2, Strelka  
- snpEff, Annovar, SIFT, pph2, Custom Annotation  
- Coverage Plot, Circos Plot, Hotspot Coverage Box Plot  
- Create input format for oncogenomics database (Patient Level)  
- Make Actionable Classification for Germline and Somatic Mutations   
- Copy number based on the simple T/N LogRatio (N cov >=30), Corrected for Total # Reads  
- Copy number, tumor purity using sequenza
- Contamination using [conpair](https://github.com/nygenome/Conpair)  
- HLA Typing  
	* [HLAminer](www.bcgsc.ca/platform/bioinfo/software/hlaminer) v1.3.1   
	* [seq2HLA](https://bitbucket.org/sebastian_boegel/seq2hla)  


### RNASeq:
----
- QC
- Broad Standard Practices on STAR bam
- RSEM pipeline
