[![Snakemake](https://img.shields.io/badge/snakemake- >=3.8.0-brightgreen.svg?style=flat-square)](https://snakemake.bitbucket.io)

## Introduction
This NGS pipeline is designed for NCI Patient-Derived Models Repository (PDMR) project (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr)

## Installation

The easiest way for installation is to clone the repository using command below.
```
git clone https://github.com/FNL-MoCha/PDMR_landscape_paper_public.git
```
This pipeline is available on NIH biowulf cluster, contact us (NCI_PDM_Repository@mail.nih.gov or mochanctn@mail.nih.gov) if you would like to do a test run.

## Scope
### Inputs:
  - Raw sequencing reads (paired-end FASTQ files)
  - Required references (hg19 reference genome fasta)
  - Target BED files


### Outputs:
- QC reports
- Aligned reads: coordinate-sorted BAM files with index and alignment metrics
- Variant calls:
	- Tumor-only: VCF file generated by HaplotypeCaller (https://gatk.broadinstitute.org/hc/en-us/articles/360037225632-HaplotypeCaller)/ Platypus (https://github.com/andyrimmer/Platypus)
	- Tumorâ€“normal pairs: VCF file generated by GATK Mutect2 (https://gatk.broadinstitute.org/hc/en-us/articles/360037593851-Mutect2)
- Copy-number profiles generated with CNVkit (https://cnvkit.readthedocs.io/en/stable/)
- Microsatellite Instability (MSI) (https://github.com/niu-lab/msisensor2)

## Prerequisites
- Linux environment (HPC or workstation)
- Bioinformatics Tools Listed in [config files](config/config_common.json)

## Documentations for PDMR NGS Pipeline
---
- Annotation of Variants from Whole Exome Assay with OncoKB (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0053.pdf)
- Whole Exome Sequencing Data Analysis Pipeline and Specifications(https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0011.pdf)
- Detection of Microsatellite Instability (MSI) Status from Whole Exome Sequencing Data (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0057.pdf)
- Individual Ancestry Estimation from
Whole Exome Sequencing Data (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0058.pdf)
- Detection of Consensus Molecular Subtype (CMS) Classification
of Colorectal Cancer from RNASeq Data (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0063.pdf)
- Detection of Human Leukocyte Antigen (HLA) Typing from Whole Exome
Sequencing Data (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0064.pdf)
- RNASeq Transciptome Data Analysis Pipeline and Specifications (https://dctd.cancer.gov/drug-discovery-development/reagents-materials/pdmr/using-pdmr/sops/mccrd-sop0012.pdf)